diff --git a/src/main/scala/ir/rtl/hardwaretype/ComplexHW.scala b/src/main/scala/ir/rtl/hardwaretype/ComplexHW.scala
index f5929d1..fe5f9de 100644
--- a/src/main/scala/ir/rtl/hardwaretype/ComplexHW.scala
+++ b/src/main/scala/ir/rtl/hardwaretype/ComplexHW.scala
@@ -37,7 +37,10 @@ case class ComplexHW[T](hw: HW[T]) extends HW[Complex[T]](hw.size * 2) (using Co
 
   override def minus(lhs: Sig[Complex[T]], rhs: Sig[Complex[T]]): Sig[Complex[T]] = Cpx(Re(lhs) - Re(rhs), Im(lhs) - Im(rhs))
 
-  override def times(lhs: Sig[Complex[T]], rhs: Sig[Complex[T]]): Sig[Complex[T]] = Cpx(Re(lhs) * Re(rhs) - Im(lhs) * Im(rhs), Re(lhs) * Im(rhs) + Im(lhs) * Re(rhs))
+  override def times(lhs: Sig[Complex[T]], rhs: Sig[Complex[T]]): Sig[Complex[T]] =
+    val z = (Re(lhs) - Im(lhs)) * Re(rhs)
+    Cpx(Im(lhs) * (Re(rhs) - Im(rhs)) + z,
+        Re(lhs) * (Re(rhs) + Im(rhs)) - z)
 
   override def bitsOf(const: Complex[T]): BigInt = (hw.bitsOf(const.im) << hw.size) + hw.bitsOf(const.re)
 
diff --git a/src/main/scala/transforms/fft/DiagC.scala b/src/main/scala/transforms/fft/DiagC.scala
index dff05cb..6bd6049 100644
--- a/src/main/scala/transforms/fft/DiagC.scala
+++ b/src/main/scala/transforms/fft/DiagC.scala
@@ -54,7 +54,7 @@ case class DiagC(override val n: Int, r: Int, l: Int) extends SPL[Complex[Double
     override def implement(inputs: Seq[Sig[Complex[Double]]]): Seq[Sig[Complex[Double]]] = (0 until K).map(p => 
       val twiddles = Vector.tabulate(T)(c => coef((c * K) + p))
       val twiddleHW = hw match
-        case ComplexHW(FixedPoint(magnitude, fractional)) => ComplexHW(FixedPoint(2, magnitude + fractional - 2))
+        case ComplexHW(FixedPoint(magnitude, fractional)) => ComplexHW(FixedPoint(2, magnitude + fractional - 6))
         case _ => hw
       val control = Timer(T)
       val twiddle = ROM(twiddles, control)(using twiddleHW)
@@ -85,7 +85,7 @@ case class StreamDiagC(override val n: Int, r: Int) extends SPL[Complex[Double]]
       val j = p % (1 << r)
       val coefs = Vector.tabulate(1 << (this.n - r))(i => DFT.omega(this.n, i * j))
       val twiddleHW = hw match
-        case ComplexHW(FixedPoint(magnitude, fractional)) => ComplexHW(FixedPoint(2, magnitude + fractional - 2))
+        case ComplexHW(FixedPoint(magnitude, fractional)) => ComplexHW(FixedPoint(2, magnitude + fractional - 6))
         case _ => hw
       val control1 = Timer(T) :: Const(p >> r)(using Unsigned(this.k - r))
       val control2a = SetCounter(this.n / r)
diff --git a/src/main/scala/transforms/fft/DiagE.scala b/src/main/scala/transforms/fft/DiagE.scala
index c7d5ded..8c0e53c 100644
--- a/src/main/scala/transforms/fft/DiagE.scala
+++ b/src/main/scala/transforms/fft/DiagE.scala
@@ -53,7 +53,7 @@ case class DiagE private (override val n: Int, r: Int, l: Int) extends SPL[Compl
     override def implement(inputs: Seq[Sig[Complex[Double]]]): Seq[Sig[Complex[Double]]] = (0 until K).map(p => 
       val twiddles = Vector.tabulate(T)(c => coef((c * K) + p))
       val twiddleHW = hw match // The hardware datatype used for the twiddles is the same as the one used by the data, EXCEPT in case of FixedPoint: to maximize precision, we store as many fractional bits as possible, as twiddles are in the unit circle. 
-        case ComplexHW(FixedPoint(magnitude, fractional)) => ComplexHW(FixedPoint(2, magnitude + fractional - 2))
+        case ComplexHW(FixedPoint(magnitude, fractional)) => ComplexHW(FixedPoint(2, magnitude + fractional - 6))
         case _ => hw
       val control = Timer(T)
       val twiddle = ROM(twiddles, control)(using twiddleHW)
